ttnn/cpp/ttnn/mlir_graph_capture_l1_interface.cpp

```diff
@@ -0,0 +1,198 @@
+#include "mlir_graph_capture_l1_interface.hpp"
+
+#include <cstdint>
+#include <tuple>
+
+#include "host_api.hpp"
+#include "impl/device/device.hpp"
+#include "third_party/json/json.hpp"
+#include "ttnn/graph/graph_operation_queries.hpp"
+#include "ttnn/operations/common/l1_interface_common.hpp"
+#include "ttnn/operations/creation.hpp"
+#include "ttnn/operations/eltwise/binary/binary.hpp"
+#include "ttnn/operations/eltwise/binary/binary_l1_interface.hpp"
+#include "ttnn/operations/eltwise/unary/unary.hpp"
+#include "ttnn/operations/eltwise/unary/unary_l1_interface.hpp"
+#include "ttnn/operations/matmul/matmul.hpp"
+#include "ttnn/operations/matmul/matmul_l1_interface.hpp"
+#include "ttnn/operations/normalization/softmax/softmax.hpp"
+#include "ttnn/operations/normalization/softmax/softmax_l1_interface.hpp"
+#include "ttnn/tensor/tensor.hpp"
+#include "ttnn/tensor/types.hpp"
+
+namespace ttnn::mlir_interface::graph_capture {
+
+class ScopedDeviceContext {
+   public:
+    ScopedDeviceContext() : m_device(tt::tt_metal::CreateDevice(0)) {}
+
+    ~ScopedDeviceContext() { tt::tt_metal::CloseDevice(m_device); }
+
+    tt::tt_metal::Device& get_device() { return *m_device; }
+
+   private:
+    tt::tt_metal::Device* m_device;
+};
+
+static ttnn::Tensor create_tensor(tt::tt_metal::Device& device, const L1InterfaceOperandParams& params) {
+    return ttnn::zeros(
+        std::get<ttnn::types::Shape>(params),
+        std::get<tt::tt_metal::DataType>(params),
+        std::get<tt::tt_metal::Layout>(params),
+        device,
+        std::get<tt::tt_metal::MemoryConfig>(params));
+}
+
+static nlohmann::json get_unary_op_trace(
+    const L1InterfaceOperandParams& input, const L1InterfaceOperandParams& output) {
+    ScopedDeviceContext ctx;
+
+    auto input_tensor = create_tensor(ctx.get_device(), input);
+
+    auto call = [&] {
+        const auto output_tensor = ttnn::relu(input_tensor, std::get<tt::tt_metal::MemoryConfig>(output));
+        return output_tensor;
+    };
+
+    return graph::query_trace(call);
+}
+
+static nlohmann::json get_binary_op_trace(
+    const L1InterfaceOperandParams& input_a,
+    const L1InterfaceOperandParams& input_b,
+    const L1InterfaceOperandParams& output) {
+    ScopedDeviceContext ctx;
+
+    auto input_tensor_a = create_tensor(ctx.get_device(), input_a);
+    auto input_tensor_b = create_tensor(ctx.get_device(), input_b);
+
+    auto call = [&] {
+        const auto output_tensor = ttnn::add(
```

Comment:
is it feasible to (long term) replace ttnn:add with string eltwise_op_name or something like that? not urgent - only to understand this solution better.

Reply:
It is, I will add a TODO in the code